File: single_number/singleNumber.py
#!/usr/local/bin/python3
#coding=utf-8

from typing import List

class SingleNumber:
    def hashTraversing(self, nums: List[int]) -> int:
        hash_map = {}
        for i in nums:
            if i in hash_map:
                hash_map[i] += 1
            else:
                hash_map[i] = 1

        for k, v in hash_map.items():
            if v == 1:
                return k
        
        raise ValueError("Can't find it")

    def xor(self, nums) -> int:
        complement = 0
        for v in nums:
            complement ^= v
        return complement

def main():
    numbers = [4,1,2,1,2]
    result = 4

    obj = SingleNumber()
    print("数组: %s, 结果: %d" % (numbers, result))
    print("HashMap 输出结果: %d" % obj.hashTraversing(numbers))
    print("异或运算(XOR) 输出结果: %d \n" % obj.xor(numbers))

if __name__ == "__main__":
    main()

File: single_number/SingleNumber.js
class SingleNumber {
    hashTraversing(nums) {
        const hash_map = {}
        for (const v of nums) {
            if (hash_map[v]) {
                hash_map[v]++
            } else {
                hash_map[v] = 1
            }
        }

        for (const k of Object.keys(hash_map)) {
            if (hash_map[k] == 1) {
                return k
            }
        }

        throw Error("Can't find solution!")
    }

    xor(nums) {
        let complement = 0
        for (const v of nums) {
            complement ^= v
        }
        return complement
    }

}

const numbers = [4, 1, 2, 1, 2]
const result = 4

const obj = new SingleNumber()
console.log("数组: %s, 结果: %d", numbers, result)
console.log("HashMap 输出结果: %d", obj.hashTraversing(numbers))
console.log("异或运算(XOR) 输出结果: %d", obj.xor(numbers))